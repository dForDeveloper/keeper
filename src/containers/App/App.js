import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withRouter, Route } from 'react-router-dom';
import NoteContainer from '../NoteContainer/NoteContainer';
import NoteForm from '../NoteForm/NoteForm';
import { fetchData } from '../../utils/api';
import { setNotes, setError } from '../../actions';

export class App extends Component {

  fetchNotes = async () => {
    const { setNotes, setError } = this.props;
    try {
      const response = await fetchData('http://localhost:3001/api/v1/notes');
      setNotes(await response.json());
    } catch (error) {
      setError(error);
    }
  }

  componentDidMount = () => {
    this.fetchNotes();
  }

  render() {
    return (
      <div className="App">
        <h1>Trapper Keeper</h1>
        <Route path='/new-note' component={NoteForm}/>
        <Route path='/' component={NoteContainer}/>
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  notes: state.notes,
  isLoading: state.isLoading,
  error: state.error
});

export const mapDispatchToProps = (dispatch) => ({
  setNotes: (notes) => dispatch(setNotes(notes)),
  setError: (error) => dispatch(setError(error))
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));

App.propTypes = {
  notes: PropTypes.array,
  isLoading: PropTypes.bool,
  error: PropTypes.string,
  setNotes: PropTypes.func,
  setError: PropTypes.func
}