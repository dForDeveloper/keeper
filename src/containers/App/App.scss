.App {
  text-align: center;
}

.App--h1 {
  background-color: #fff;
  color: $secondary-dark;
  font-size: 1.6rem;
  letter-spacing: 3px;
  font-weight: 500;
  margin: 0;
  padding: 0% 4%;
  height: 65px;
  display: flex;
  align-items: center;
  justify-content: left;
  box-shadow: $heavy-shadow;
  z-index: 10;
  position: relative;
}

.App--notebook-icon {
  height: 1.7rem;
  margin-right: 4px;
}