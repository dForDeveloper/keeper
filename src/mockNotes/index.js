export const mockNotes = [
  { 
    id: 'ijf', 
    title: 'Learn Redux', 
    listItems: [
      { 
        id: 'asg', 
        description: 'read through lesson', 
        isComplete: false
      }, 
      { 
        id: 'rhf', 
        description: 'follow along with tutorial', 
        isComplete: false
      }
    ]
  },
  { 
    id: 'pdo', 
    title: 'Start portfolio', 
    listItems: [
      { 
        id: 'cat', 
        description: 'Setup user account', 
        isComplete: false
      }
    ]
  },
  { 
    id: 'bah', 
    title: 'Grocery shopping', 
    listItems: [
      { 
        id: 'rey', 
        description: 'Stop by King Soopers', 
        isComplete: false
      }, 
      { 
        id: 'moo', 
        description: 'Stop by Sprouts', 
        isComplete: false
      }
    ]
  },
  { 
    id: 'fah', 
    title: 'Make dinner', 
    listItems: [
      { 
        id: 'ley', 
        description: 'gather ingredients', 
        isComplete: false
      }, 
      { 
        id: 'mei', 
        description: 'preheat oven', 
        isComplete: false
      }
    ]
  }
]

export const mockNote = {
  id: 'rzz',
  title: 'Workout',
  listItems: [
    {
      id: 'lpo',
      description: '20 situps',
      isComplete: false
    },
    {
      id: 'lqw',
      description: '15 pushups',
      isComplete: false
    }
  ]
}