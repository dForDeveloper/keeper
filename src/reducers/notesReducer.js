export const notesReducer = (state = [], action) => {
  switch (action.type) {
    case 'SET_NOTES':
      return action.notes;
    case 'ADD_NOTE':
      return [...state, action.note];
    case 'UPDATE_NOTE':
      return state.map(note => {
        const { id } = action.note;
        return note.id === id ? action.note : note;
      });
    case 'DELETE_NOTE':
      return state.filter(note => note.id !== action.id);
    default:
      return state;
  }
}