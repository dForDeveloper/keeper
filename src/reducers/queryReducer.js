export const queryReducer = (state = '', action) => {
  switch (action.type) {
    case 'SET_QUERY':
      return action.query;
    default:
      return state;
  }
}