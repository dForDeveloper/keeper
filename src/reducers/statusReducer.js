export const statusReducer = (state = 0, action) => {
  switch (action.type) {
    case 'SET_STATUS':
      return action.code;
    default:
      return state;
  }
}