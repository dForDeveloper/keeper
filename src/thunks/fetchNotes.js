import { setNotes, setError } from '../actions';
import { fetchData } from '../utils/api';

export const fetchNotes = () => {
  return async (dispatch) => {
    try {
      const response = await fetchData('http://localhost:3001/api/v1/notes');
      dispatch(setNotes(await response.json()));
    } catch (error) {
      dispatch(setError(error));
    }
  }
}