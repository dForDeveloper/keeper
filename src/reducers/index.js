import { combineReducers } from 'redux';
import { notesReducer } from './notesReducer';
import { isLoadingReducer } from './isLoadingReducer';
import { errorReducer } from './errorReducer';

export const rootReducer = combineReducers({
  notes: notesReducer,
  isLoading: isLoadingReducer,
  error: errorReducer
});