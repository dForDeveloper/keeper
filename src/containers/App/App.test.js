import React from 'react';
import ReactDOM from 'react-dom';
import { App, mapStateToProps, mapDispatchToProps} from './App';
import { fetchNotes } from '../../thunks/fetchNotes';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { rootReducer } from '../../reducers/';
import { createStore } from 'redux';
import { shallow } from 'enzyme';
import { mockNotes } from '../../mockNotes';

const mockProps = {
  notes: mockNotes,
  isLoading: false,
  error: '',
  fetchNotes: jest.fn(() => true)
}

const mockMatch = { params: { id: 'ijf' }, path: 'notes/ijf' };

jest.mock('../../thunks/fetchNotes.js');

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    const store = createStore(rootReducer);
    ReactDOM.render(<Provider store={store}><BrowserRouter><App {...mockProps}/></BrowserRouter></Provider>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  describe('App component', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<App {...mockProps}/>);
    });
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call fetchNotes on componentDidMount', () => {
      wrapper.instance().componentDidMount();
      expect(mockProps.fetchNotes).toHaveBeenCalled();
    });

    it('should return an array of JSX when there is a note', () => {
      const result = wrapper.instance().getNotesRoute({ match: mockMatch });
      expect(result).toHaveLength(2);
    });

    it('should return an Error404 component when there is no note', () => {
      const mockMatch = { params: { id: 'hghgfjhg' }, path: 'notes/hghgfjhg' };
      const result = wrapper.instance().getNotesRoute({ match: mockMatch });
      const wrapper404 = shallow(result);
      expect(wrapper404.find('.Error404')).toHaveLength(1);
    });

    it('should return an array of JSX', () => {
      const result = wrapper.instance().getNewNoteRoute({ match: mockMatch });
      expect(result).toHaveLength(2);
    });
  });

  describe('mapStateToProps', () => {
    it('should return a props object with notes, isLoading and an error', () => {
      const mockState = {
        notes: mockNotes,
        isLoading: false,
        error: '',
        status: 0
      };
      const expected = {
        notes: mockNotes,
        isLoading: false,
        error: ''
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with a fetchNotes thunk', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchNotes();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchNotes();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

});
