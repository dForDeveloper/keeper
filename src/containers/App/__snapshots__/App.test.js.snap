// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App component should match the snapshot 1`] = `
<div
  className="App"
>
  <h1
    className="App--h1"
  >
    <img
      alt="note icon"
      className="App--notebook-icon"
      src="notebook.svg"
    />
    Keeper
  </h1>
  <Switch>
    <Route
      path="/new-note"
      render={[Function]}
    />
    <Route
      path="/notes/:id"
      render={[Function]}
    />
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
    />
  </Switch>
</div>
`;

exports[`App App component should match the snapshot when isLoading is true 1`] = `
<div
  className="App"
>
  <h1
    className="App--h1"
  >
    <img
      alt="note icon"
      className="App--notebook-icon"
      src="notebook.svg"
    />
    Keeper
  </h1>
  <Loader />
  <Switch>
    <Route
      path="/new-note"
      render={[Function]}
    />
    <Route
      path="/notes/:id"
      render={[Function]}
    />
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
    />
  </Switch>
</div>
`;
